src/coreclr/nativeaot/Runtime/riscv64/UniversalTransition.S

```diff
@@ -84,88 +86,93 @@
-        fld f0, 0x100(sp)         // Restore f0
-        fld f1, 0x108(sp)         // Restore f1
-        fld f2, 0x110(sp)         // Restore f2
-        fld f3, 0x118(sp)         // Restore f3
-        fld f4, 0x120(sp)         // Restore f4
-        fld f5, 0x128(sp)         // Restore f5
-        fld f6, 0x130(sp)         // Restore f6
-        fld f7, 0x138(sp)         // Restore f7
-
-        // Restore argument registers
-        ld a0, 0x140(sp)         // Restore a0
-        ld a1, 0x148(sp)         // Restore a1
-        ld a2, 0x150(sp)         // Restore a2
-        ld a3, 0x158(sp)         // Restore a3
-        ld a4, 0x160(sp)         // Restore a4
-        ld a5, 0x168(sp)         // Restore a5
-        ld a6, 0x170(sp)         // Restore a6
-        ld a7, 0x178(sp)         // Restore a7
-
-        // Restore FP and RA registers, and free the allocated stack block
-        EPILOG_RESTORE_REG_PAIR_INDEXED  fp, ra, STACK_SIZE
-
-        // Tailcall to the target address in t2
-        jalr t2, 0
+        ld a2, 0x10(a1)
+        ld a3, 0x18(a1)
+        ld a4, 0x20(a1)
+        ld a5, 0x28(a1)
+        ld a6, 0x30(a1)
+        ld a7, 0x38(a1)
+#endif // TRASH_SAVED_ARGUMENT_REGISTERS
+
+        add         a0, sp, DISTANCE_FROM_CHILDSP_TO_RETURN_BLOCK  # First parameter to target function is a pointer to the return block
+        mv          a1, t1                                         # Second parameter to target function
+        jalr        ra, t0, 0
+
+ALTERNATE_ENTRY ReturnFrom\FunctionName
+
+        # Move the result (the target address) to a2 so it doesn't get overridden when we restore the
+        # argument registers.
+        mv          a2, a0
+
+        # Restore floating point registers
+        fld         fa0, FLOAT_ARG_OFFSET(sp)
+        fld         fa1, FLOAT_ARG_OFFSET + 0x08(sp)
+        fld         fa2, FLOAT_ARG_OFFSET + 0x10(sp)
+        fld         fa3, FLOAT_ARG_OFFSET + 0x18(sp)
+        fld         fa4, FLOAT_ARG_OFFSET + 0x20(sp)
+        fld         fa5, FLOAT_ARG_OFFSET + 0x28(sp)
+        fld         fa6, FLOAT_ARG_OFFSET + 0x30(sp)
+        fld         fa7, FLOAT_ARG_OFFSET + 0x38(sp)
+
+        # Restore the argument registers
+        ld          a0, ARGUMENT_REGISTERS_OFFSET(sp)
+        ld          a1, ARGUMENT_REGISTERS_OFFSET + 0x08(sp)
+        ld          a2, ARGUMENT_REGISTERS_OFFSET + 0x10(sp)
+        ld          a3, ARGUMENT_REGISTERS_OFFSET + 0x18(sp)
+        ld          a4, ARGUMENT_REGISTERS_OFFSET + 0x20(sp)
+        ld          a5, ARGUMENT_REGISTERS_OFFSET + 0x28(sp)
+        ld          a6, ARGUMENT_REGISTERS_OFFSET + 0x30(sp)
+        ld          a7, ARGUMENT_REGISTERS_OFFSET + 0x38(sp)
+
+        # Restore FP and RA registers, and free the allocated stack block
+        ld          s0, STACK_SIZE - 8(sp)
+        ld          ra, STACK_SIZE - 16(sp)
+        addi        sp, sp, STACK_SIZE
+
+        # Tailcall to the target address.
+        jalr        zero, a2, 0
```

Comment:
```suggestion
        jr          a2
```
Use `jr` pseudo code instead of the degenerated form to make this more readable?